core/domain/state_domain.py

```diff
@@ -916,8 +940,917 @@ def get_all_outcomes(self) -> List[Outcome]:
+        range_var['ub_inclusive'] = ub_inclusive
+
+    def _is_enclosed_by(
+        self,
+        test_range: RangeVariableDict,
+        base_range: RangeVariableDict
+    ) -> bool:
+        """Returns `True` when `test_range` variable lies within
+        `base_range` variable.
+
+        Args:
+            test_range: RangeVariableDictDict. It represents the variable for
+                which we have to check the range.
+            base_range: RangeVariableDictDict. It is the variable to which
+                the range is compared.
+
+        Returns:
+            bool. Returns True if test_range lies
+            within base_range.
+        """
+        if (
+            base_range['lower_bound'] is None or
+            test_range['lower_bound'] is None or
+            base_range['upper_bound'] is None or
+            test_range['upper_bound'] is None
+        ):
+            return False
+        lb_satisfied = (
+            base_range['lower_bound'] < test_range['lower_bound'] or 
+            (
+                base_range['lower_bound'] == test_range['lower_bound'] and
+                (not test_range['lb_inclusive'] or base_range['lb_inclusive'])
+            )
+        )
+        ub_satisfied = (
+            base_range['upper_bound'] > test_range['upper_bound'] or 
+            (
+                base_range['upper_bound'] == test_range['upper_bound'] and
+                (not test_range['ub_inclusive'] or base_range['ub_inclusive'])
+            )
+        )
+        return lb_satisfied and ub_satisfied
+
+    def _should_check_range_criteria(
+        self,
+        earlier_rule: RuleSpec,
+        later_rule: RuleSpec
+    ) -> bool:
+        """Compares the rule types of two rule specs to determine whether
+        to check for range enclosure.
+
+        Args:
+            earlier_rule: RuleSpec. Previous rule.
+            later_rule: RuleSpec. Current rule.
+
+        Returns:
+            bool. Returns True if the rules passes the range criteria check.
+        """
+        if earlier_rule.rule_type in (
+            'HasDenominatorEqualTo', 'IsEquivalentTo', 'IsLessThan'
+            'IsEquivalentToAndInSimplestForm', 'IsGreaterThan'
+        ):
+            return True
+        return later_rule.rule_type in (
+            'HasDenominatorEqualTo', 'IsLessThan', 'IsGreaterThan'
+        )
+
+    def _get_rule_value_of_fraction_interaction(
+        self,
+        rule_spec: RuleSpec
+    ) -> float:
+        """Returns rule value of the rule_spec of FractionInput interaction so
+        that we can keep track of rule's range.
+
+        Args:
+            rule_spec: RuleSpec. Rule spec of an answer group.
+
+        Returns:
+            rule_value_f: float. The value of the rule spec.
+        """
+        rule_value_f = rule_spec.inputs['f']
+        value: float = (
+            rule_value_f['wholeNumber'] +
+            float(rule_value_f['numerator']) / rule_value_f['denominator']
+        )
+
+        return value
```

Comment:
```suggestion
        return (
            rule_value_f['wholeNumber'] +
            float(rule_value_f['numerator']) / rule_value_f['denominator']
        )
```
Not sure if it is possible.

Reply:
No, I need to do this for the mypy